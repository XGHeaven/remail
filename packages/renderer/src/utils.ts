/**
 * Generate full permtation array.
 *
 * @example
 * const perms = Array.from(permutation({
 *   a: [1, 2],
 *   b: ['A', 'B', 'C']
 * }))
 * perms === [{a: 1, b: 'A'}, {a: 2, b: 'A'}, {a: 1, b: 'B'}, ... etc] // length is 6
 */
export function *permutation<T extends {}>(permGroup: {
  [K in keyof T]: T[K][]
}): Generator<T, void, void> {
  const keys = Object.keys(permGroup) as (keyof T)[]
  const counts = keys.map(key => permGroup[key].length)
  const total = counts.reduce((t, n) => t * n, 1)

  for (let i = 0; i < total; i++) {
    const ret = {} as T
    let num = i

    for (let j = 0; j < keys.length; j++) {
      const key = keys[j]
      const count = counts[j]
      ret[key] = permGroup[key][num % count]
      num = Math.floor(num / count)
    }

    yield ret
  }
}
