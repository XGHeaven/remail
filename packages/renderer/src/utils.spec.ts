import { permutation } from './utils'

describe('permutation', () => {
  it('should works', () => {
    const perms = Array.from(permutation(
      {
        a: [1, 2, 3],
        b: ['A', 'B'],
        c: ['a', 'b'],
      }
    ))
    expect(perms).toHaveLength(12)
    expect(
      perms
    ).toMatchSnapshot()
  })
})
