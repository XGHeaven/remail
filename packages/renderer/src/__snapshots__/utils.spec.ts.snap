// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`permutation should works 1`] = `
Array [
  Object {
    "a": 1,
    "b": "A",
    "c": "a",
  },
  Object {
    "a": 2,
    "b": "A",
    "c": "a",
  },
  Object {
    "a": 3,
    "b": "A",
    "c": "a",
  },
  Object {
    "a": 1,
    "b": "B",
    "c": "a",
  },
  Object {
    "a": 2,
    "b": "B",
    "c": "a",
  },
  Object {
    "a": 3,
    "b": "B",
    "c": "a",
  },
  Object {
    "a": 1,
    "b": "A",
    "c": "b",
  },
  Object {
    "a": 2,
    "b": "A",
    "c": "b",
  },
  Object {
    "a": 3,
    "b": "A",
    "c": "b",
  },
  Object {
    "a": 1,
    "b": "B",
    "c": "b",
  },
  Object {
    "a": 2,
    "b": "B",
    "c": "b",
  },
  Object {
    "a": 3,
    "b": "B",
    "c": "b",
  },
]
`;
